File: Algo/Leetcode/209MinSizeSubArraySum.cpp
// two pointer(sliding window)
class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int l = 0; int r = 0;
        int sum = 0;
        int min_len = nums.size() + 1;
        while(r < nums.size()){
            sum += nums[r++];
            while(sum >= target){
                min_len = min(min_len, r-l);
                sum -= nums[l++];
            }
        }
        return min_len == nums.size() + 1? 0:min_len;
    }
};

File: Algo/Leetcode/209MinSizeSubArraySum.py
class Solution(object):
    def minSubArrayLen(self, target, nums):
        """
        :type target: int
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        min_len = n + 1
        l, r = 0, 0
        s = 0
        while l < n and r < n:
            s += nums[r]
            while s >= target and l < n:        # you should keep the while loop
                min_len = min(min_len, r - l + 1)
                s -= nums[l]
                l += 1

            r += 1

        return min_len if min_len != n+1 else 0